Algorithm.Framework/Portfolio/BlackLittermanPortfolioConstructionModel.py

```diff
@@ -0,0 +1,233 @@
+from datetime import timedelta
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize
+from itertools import groupby
+from numpy import dot, transpose
+from numpy.linalg import inv
+
+### <summary>
+### Provides an implementation of Black-Litterman portfolio optimization. The model adjusts equilibrium market 
+### returns by incorporating views from multiple alpha models and therefore to get the optimal risky portfolio 
+### reflecting those views. If insights of all alpha models have None magnitude or there are linearly dependent 
+### vectors in link matrix of views, the expected return would be the implied excess equilibrium return. 
+### The interval of weights in optimization method can be changed based on the long-short algorithm.
+### The default model uses the 0.0025 as weight-on-views scalar parameter tau. The optimization method 
+### maximizes the Sharpe ratio with the weight range from -1 to 1.
+### </summary>
+class BlackLittermanPortfolioConstructionModel:
+    def __init__(self, *args, **kwargs):
+        """Initialize the model
+        Args:
+            lookback(int): Historical return lookback period
+            resolution: The resolution of the history price 
+            minimum_weight(float): The lower bounds on portfolio weights
+            maximum_weight(float): The upper bounds on portfolio weights
+            risk_free_rate(float): The risk free rate
+            self.tau(float): The model parameter indicating the uncertainty of the CAPM prior """
+            
+        self.lookback = kwargs['lookback'] if 'lookback' in kwargs else 5
+        self.resolution = kwargs['resolution'] if 'resolution' in kwargs else Resolution.Daily
+        self.minimum_weight = kwargs['minimum_weight'] if 'minimum_weight' in kwargs else -1
+        self.maximum_weight = kwargs['maximum_weight'] if 'maximum_weight' in kwargs else 1
+        self.risk_free_rate = kwargs['risk_free_rate'] if 'risk_free_rate' in kwargs else 0
+        self.tau = kwargs['tau'] if 'tau' in kwargs else 0.025
+        self.symbolDataDict = {}
+        self.pendingRemoval = []
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm: The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        
+        for symbol in self.pendingRemoval:
+            yield PortfolioTarget.Percent(algorithm, symbol, 0)
+        self.pendingRemoval.clear()
+        
+        # construct the dataframe of historical return
+        price = {}
+        for symbol, data in self.symbolDataDict.items():
+            price[str(symbol)] = data.PriceSeries()
+        df_price = pd.DataFrame(price)[::-1]
+        algorithm.Log(str(df_price))
+        returns = df_price.pct_change().dropna()
```

Comment:
I think we should have a return series instead of a price series in `SymbolData`. It would save us a `pandas.pct_change().dropna()` calls.
However, it is a good example to show a different approach of what we have done with the `MeanVarianceOptimizationPortfolioConstructionModel`.